packages/grafana-data/src/utils/QueryEditorSanitizer.tsx

```diff
@@ -0,0 +1,30 @@
+import { ComponentType, useEffect } from 'react';
+
+import { DataQuery, DataSourceJsonData } from '@grafana/schema';
+
+import { DataSourceApi, DataSourceOptionsType, DataSourceQueryType, QueryEditorProps } from '../types/datasource';
+
+// QueryEditorSanitizer is a higher order component that can be used to wrap a QueryEditor component
+// and ensure that the query is sanitized before being passed to the QueryEditor.
+export function QueryEditorSanitizer<
+  DSType extends DataSourceApi<TQuery, TOptions>,
+  TQuery extends DataQuery = DataSourceQueryType<DSType>,
+  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>,
+>(QueryEditor: ComponentType<QueryEditorProps<DSType, TQuery, TOptions>>) {
+  const WithExtra = (props: QueryEditorProps<DSType, TQuery, TOptions>) => {
+    useEffect(() => {
+      if (props.query && props.datasource.migrateQuery) {
+        const migrated = props.datasource.migrateQuery(props.query);
+        if (migrated instanceof Promise) {
+          migrated.then((migrated) => {
+            props.onChange(migrated);
+          });
+        } else {
+          props.onChange(migrated);
+        }
+      }
+    }, []); // eslint-disable-line react-hooks/exhaustive-deps
+    return <QueryEditor {...props} />;
+  };
+  return WithExtra;
```

Comment:
Just trying to understand the flow here. Let's say the plugin author has written a migration that moves a prop in the query object.
Before: 
```ts
{
 refId: 'A',
 sql: '...'
}
```
After: 
```ts
{
 refId: 'A',
 query: {
    sql: '123'
 },
}
```

Typescript types are updated and query editor changed according to new structure. In that case the plugin query editor would/could crash if the legacy query structured is passed as props during first render. So if a migration is defined, we need to protect the plugin editor from receiving the legacy structure until the migration is completed? 

If a migration is defined, maybe we can render a [skeleton](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/utils/skeleton.tsx) meanwhile migration is happening? Alternatively, we render the actual editor but with default query (datasource.getDefaultQuery) instead of the legacy query. Thoughts?

Reply:
That's right. I wanted to avoid having to deal with state and all that but it'd be more consistent if `QueryEditor` is only render when fully migrated.